File: backend/src/main/kotlin/com/trackwatch/backend/clients/spotify/SpotifyAuthApiClient.kt
package com.trackwatch.backend.clients.spotify

import com.trackwatch.backend.clients.spotify.dto.SpotifyTokenDTO
import com.trackwatch.backend.exception.*
import com.trackwatch.backend.exception.ErrorCode
import com.trackwatch.backend.utils.service.MetricService
import org.slf4j.LoggerFactory

import org.springframework.stereotype.Component
import org.springframework.util.LinkedMultiValueMap
import org.springframework.util.MultiValueMap
import java.util.Base64

@Component
class SpotifyAuthApiClient(metricService: MetricService): SpotifyApiClient(metricService) {

    private val log = LoggerFactory.getLogger(SpotifyAuthApiClient::class.java)

    fun refreshAccessTokenWithRetries(refreshToken: String, maxAttempts: Int = 3): SpotifyTokenDTO {
        var lastException: Exception? = null

        for (attempt in 1..maxAttempts) {
            try {
                return refreshAccessToken(refreshToken)
            } catch (e: Exception) {
                lastException = e
                log.warn("Spotify token refresh failed (attempt $attempt/$maxAttempts): $e")

                if (attempt < maxAttempts) {
                    val waitTimeMs = 1000L * attempt // Exponential backoff
                    log.info("Retrying in ${waitTimeMs}ms")

                    try {
                        Thread.sleep(waitTimeMs)
                    } catch (ie: InterruptedException) {
                        Thread.currentThread().interrupt()
                        throw ie
                    }
                }
            }
        }

        // If all attempts fail, throw the last exception
        throw lastException ?: InternalServerErrorException(
            ErrorCode.UNHANDLED_EXCEPTION,
            "Failed to refresh Spotify access token after $maxAttempts attempts",
            "Unknown error"
        )
    }

    private fun refreshAccessToken(refreshToken: String): SpotifyTokenDTO {
        val authString = getAuthString()

        val bodyFormEncoded: MultiValueMap<String, String> = LinkedMultiValueMap()
        bodyFormEncoded.add("grant_type", "refresh_token")
        bodyFormEncoded.add("refresh_token", refreshToken)

        try {
            sendMetricApiCall("refreshAccessToken")

            val response = authWebClient.post()
                .uri("/token")
                .header("Authorization", "Basic $authString")
                .header("Content-Type", "application/x-www-form-urlencoded")
                .bodyValue(bodyFormEncoded)
                .retrieve()
                .bodyToMono(Map::class.java)
                .block() ?: throw InternalServerErrorException(ErrorCode.UNHANDLED_EXCEPTION, "Error while calling Spotify API", "Response is null")

            return mapToSpotifyTokenDTO(response, refreshToken)

        } catch (e: Exception) {
            throw InternalServerErrorException(ErrorCode.UNHANDLED_EXCEPTION, "Error while calling Spotify API", e.toString())
        }
    }

    private fun getAuthString(): String {
        val clientId = System.getenv("SPOTIFY_CLIENT_ID")
        val clientSecret = System.getenv("SPOTIFY_CLIENT_SECRET")

        val authString = "$clientId:$clientSecret"
        val encodedAuthString = Base64.getEncoder().encodeToString(authString.toByteArray())

        return encodedAuthString
    }

    private fun mapToSpotifyTokenDTO(response: Map<*, *>, refreshToken: String): SpotifyTokenDTO {
        return SpotifyTokenDTO(
            accessToken = response["access_token"] as String,
            refreshToken = refreshToken
        )
    }

}

File: backend/app/clients/spotify/spotify_auth_api_client.py
import time
from decouple import config
import requests
from .spotify_api_client import spotify_auth_request
from app.exceptions import InternalServerErrorException, ErrorCode, SpotifyReauthorizationRequiredException

INVALID_GRANT_ERROR = "invalid_grant"


def _spotify_error_code(exception):
  response = getattr(exception, "response", None)
  if response is None:
    return None

  try:
    return response.json().get("error")
  except ValueError:
    return None


def _is_invalid_grant(exception):
  return _spotify_error_code(exception) == INVALID_GRANT_ERROR


def _exception_details(exception):
  return getattr(exception, "details", str(exception))

def refresh_access_token_with_retries(refresh_token: str, max_attempts: int = 3):
  last_exception = None

  for attempt in range(1, max_attempts + 1):
    try: return refresh_access_token(refresh_token)
    except SpotifyReauthorizationRequiredException:
      raise
    except Exception as e:
      last_exception = e
      print(f"Spotify token refresh failed (attempt {attempt}/{max_attempts}): {_exception_details(e)}")
      if attempt < max_attempts:
        wait_time = 1 * attempt
        print(f"Retrying in {wait_time}s")
        time.sleep(wait_time)

  # If all attempts fail, throw the last exception
  if last_exception: raise last_exception
  else:
    raise InternalServerErrorException(
      ErrorCode.UNHANDLED_EXCEPTION,
      f"Failed to refresh Spotify access token after {max_attempts} attempts",
      "Unknown error"
    )

def refresh_access_token(refresh_token: str):
  client_id = config("SPOTIFY_CLIENT_ID")
  client_secret = config("SPOTIFY_CLIENT_SECRET")

  data = {
    "grant_type": "refresh_token",
    "refresh_token": refresh_token
  }

  try:
    resp = spotify_auth_request(
      method="POST",
      endpoint="/token",
      data=data,
      auth=(client_id, client_secret)
    )
    return map_to_spotify_token_dto(resp, refresh_token)

  except requests.HTTPError as e:
    if _is_invalid_grant(e):
      raise SpotifyReauthorizationRequiredException(
        ErrorCode.SPOTIFY_REAUTH_REQUIRED,
        details="Spotify refresh token is expired or invalid",
      )

    raise InternalServerErrorException(
      ErrorCode.UNHANDLED_EXCEPTION,
      "Error while calling Spotify API",
      str(e)
    )

  except Exception as e:
    raise InternalServerErrorException(
      ErrorCode.UNHANDLED_EXCEPTION,
      "Error while calling Spotify API",
      str(e)
    )

def map_to_spotify_token_dto(response: dict, refresh_token: str):
  return {
    "access_token": response.get("access_token"),
    "refresh_token": response.get("refresh_token") or refresh_token
  }
